File: backend/app/services/scheduler.py
import asyncio
import time
from typing import Optional

from app.db import atlassian_repo, jobs_repo
from app.services.jobs import enqueue_job
from app.utils.time import parse_iso_to_epoch
from app.websocket.manager import manager

POLL_INTERVAL_SEC = 600
MEMORY_PROCESS_INTERVAL_SEC = 3600  # 1 hour

_scheduler_task: Optional[asyncio.Task] = None
_memory_scheduler_task: Optional[asyncio.Task] = None
_last_memory_process_time: float = 0


def _is_due(last_sync_at: Optional[str]) -> bool:
    if not last_sync_at:
        return True
    try:
        last_ts = parse_iso_to_epoch(last_sync_at)
    except ValueError:
        return True
    return (time.time() - last_ts) >= POLL_INTERVAL_SEC


async def _queue_sync_jobs() -> None:
    for service in ("jira", "confluence"):
        accounts = await atlassian_repo.list_accounts(service=service, public=False)
        for account in accounts:
            if not _is_due(account.get("last_sync_at")):
                continue
            job_id = await jobs_repo.create_job(
                f"connector.{service}.sync", {"account_id": account["account_id"]}
            )
            await enqueue_job(job_id)
            await manager.emit_log(
                "info", f"{service} sync queued {account['account_id']}"
            )


async def _scheduler_loop() -> None:
    await manager.emit_log("info", "atlassian scheduler started")
    while True:
        try:
            await _queue_sync_jobs()
        except Exception as exc:
            await manager.emit_log("error", f"atlassian scheduler error: {exc}")
        await asyncio.sleep(POLL_INTERVAL_SEC)


async def _queue_memory_jobs() -> None:
    """Queue memory processing jobs."""
    global _last_memory_process_time

    current_time = time.time()
    if current_time - _last_memory_process_time < MEMORY_PROCESS_INTERVAL_SEC:
        return

    _last_memory_process_time = current_time

    # Process events from last hour
    job_id = await jobs_repo.create_job(
        "memory.process_events",
        {"window_minutes": 60},
    )
    await enqueue_job(job_id)
    await manager.emit_log("info", f"memory process job queued: {job_id}")

    # Sync to ChromaDB
    chroma_job_id = await jobs_repo.create_job(
        "memory.chroma_sync",
        {},
    )
    await enqueue_job(chroma_job_id)
    await manager.emit_log("info", f"chroma sync job queued: {chroma_job_id}")


async def _memory_scheduler_loop() -> None:
    """Memory processing scheduler loop."""
    await manager.emit_log("info", "memory scheduler started")
    while True:
        try:
            await _queue_memory_jobs()
        except Exception as exc:
            await manager.emit_log("error", f"memory scheduler error: {exc}")
        await asyncio.sleep(POLL_INTERVAL_SEC)


async def start_scheduler() -> None:
    global _scheduler_task, _memory_scheduler_task
    if _scheduler_task and not _scheduler_task.done():
        pass
    else:
        _scheduler_task = asyncio.create_task(_scheduler_loop())

    if _memory_scheduler_task and not _memory_scheduler_task.done():
        pass
    else:
        _memory_scheduler_task = asyncio.create_task(_memory_scheduler_loop())

File: backend-bun/src/services/scheduler.ts
/**
 * Background scheduler for automatic sync jobs
 * Polls Atlassian accounts every 10 minutes and enqueues sync jobs if due
 */

import * as atlassianRepo from "../db/atlassian.ts";
import * as jobsRepo from "../db/jobs.ts";
import { enqueueJob } from "./jobs.ts";
import { manager } from "../websocket/manager.ts";

// Poll interval: 10 minutes
const POLL_INTERVAL_MS = 10 * 60 * 1000;

let schedulerRunning = false;
let schedulerTimeout: Timer | null = null;

/**
 * Check if an account is due for sync based on last sync time
 */
function isDue(connector: string): boolean {
  const syncState = atlassianRepo.fetchSyncState(connector);

  // No previous sync - due immediately
  if (!syncState?.last_sync_at) {
    return true;
  }

  const lastSyncTs = new Date(syncState.last_sync_at).getTime();
  const elapsed = Date.now() - lastSyncTs;

  return elapsed >= POLL_INTERVAL_MS;
}

/**
 * Queue sync jobs for all due accounts
 */
async function runSchedulerCycle(): Promise<void> {
  const services: Array<"jira" | "confluence"> = ["jira", "confluence"];

  for (const service of services) {
    const accounts = atlassianRepo.fetchAtlassianAccounts(service);

    for (const account of accounts) {
      const connector = `${service}:${account.account_id}`;

      if (isDue(connector)) {
        const jobType = `connector.${service}.sync`;
        const payload = { account_id: account.account_id };

        const jobId = jobsRepo.createJob(jobType, payload);
        enqueueJob(jobId);

        manager.emitLog("info", `Scheduler: queued ${service} sync for ${account.org}`);
      }
    }
  }
}

/**
 * Main scheduler loop
 */
async function schedulerLoop(): Promise<void> {
  while (schedulerRunning) {
    try {
      await runSchedulerCycle();
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : String(error);
      manager.emitLog("error", `Scheduler error: ${errorMessage}`);
    }

    // Wait for next cycle
    await new Promise<void>((resolve) => {
      schedulerTimeout = setTimeout(resolve, POLL_INTERVAL_MS);
    });
  }
}

/**
 * Start the background scheduler
 */
export function startScheduler(): void {
  if (schedulerRunning) {
    console.log("Scheduler already running");
    return;
  }

  schedulerRunning = true;
  console.log(`Scheduler started (interval: ${POLL_INTERVAL_MS / 1000}s)`);

  // Start async loop (don't await)
  schedulerLoop();
}

/**
 * Stop the background scheduler
 */
export function stopScheduler(): void {
  schedulerRunning = false;

  if (schedulerTimeout) {
    clearTimeout(schedulerTimeout);
    schedulerTimeout = null;
  }

  console.log("Scheduler stopped");
}

/**
 * Get scheduler status
 */
export function getSchedulerStatus(): { running: boolean; interval_ms: number } {
  return {
    running: schedulerRunning,
    interval_ms: POLL_INTERVAL_MS,
  };
}
